CBMC/JBMC --cover: only store traces with --trace to avoid memory exhaustion

tautschnig opened the pull request:
Not everyone needs test inputs, especially when programs lack INPUT
instructions anyway. This helped me avoid running out of memory on a
768GB system for a particular verification problem. The same problem can
now be run with just 75GB of memory.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
CBMC --cover: only store traces with --show-test-suite

Not everyone needs test inputs, especially when programs lack INPUT
instructions anyway. Test inputs are found via goto traces, and
computing those is expensive in both time and memory required. On one
benchmark, not building goto traces reduced the overall time from 36000
seconds to 2800 seconds (a saving of 92%).

tautschnig committed:
goto trace storage: use a list instead of a vector

goto traces consume a lot of memory, and re-allocating them for each
incremental addition of a goto trace does not seem to be a good
approach. This is at the expense of more costly indexed access, which
now requires iterator increments.

peterschrammel reviewed (approved):
Good workaround in the meanwhile. Storing traces and inputs should be decoupled, though, I think.
Besides, I should probably revive the PR that enables trace output as traces are produced rather than storing them at all (but that's quite a UI change which needs to be phased in properly).

hannes-steffenhagen-diffblue commented on regression/cbmc-cover/branch-loop1/test.desc:
```diff
@@ -1,6 +1,6 @@
 CORE
 main.c
---xml-ui --cover branch
+--xml-ui --cover branch --trace
```
Seems like this is broken somehow?

tautschnig replied:
Yes, it was broken as the code generating XML output is used for both trace and test-suite output. I have therefore changed the patch to introduce a new command-line option "show-test-suite."

martin-cs reviewed (commented):
Given the performance numbers this sounds very much like the kind of thing we want and I am in favour of not storing things we don't need.  But I am a little confused over the use-case for this. `--cover` creates a bunch of assertions for the relevant coverage metric and then solves for all of them.  Then, orthogonally(? is this what @peterschrammel is suggesting?) we could output for each location:

A. reached / unreachable.
B. inputs that would cause the relevant path to be taken.
C. the execution trace that reaches each one.

So this is only fro case C?  Are there users of this?  Also, can't we output these as we compute them?  I'm not sure I see why they need to be stored at all.

tautschnig committed:
Enable merge_ireps for goto traces

A goto trace includes freshly constructed expressions, which thus lack
sharing. This results in excessive memory use when storing multiple
traces, as is done for coverage computation. On a particular benchmark,
coverage computation previously ran out of memory at 768 GB. This same
benchmark can now be run with just 80 GB of memory.

tautschnig commented:
@peterschrammel @martin-cs I'd appreciate you taking another look at this for I've made substantial changes over the version you previously reviewed/approved:

1. I've introduced `merge_irept` to effectively compress the amount of memory consumed by a trace. This alone would have solved my out-of-memory problem as I've tested, but it seems the computational cost of building all the traces is still substantial (and often unnecessary).
2. Instead of using (abusing?) the "trace" command-line option, I've instead added a new option "show-test-suite."

It may still be desirable to print test inputs as-you-go, but that's not something on my priority queue (as I don't care about those inputs values at the moment).

martin-cs commented:
@tautschnig Apologies for the lag; first week of term.  I'm guessing by the merge that you are happy with this.
